invenio_accounts/testutils.py

```diff
@@ -0,0 +1,88 @@
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+"""Invenio-Accounts utility functions for tests and testing purposes."""
+from __future__ import absolute_import, print_function
+
+import flask
+
+from flask import current_app
+
+import flask_login
+
+from flask_security import url_for_security
+from flask_security.utils import encrypt_password
+
+from werkzeug.local import LocalProxy
+
+
+# "Convenient references" (lifted from flask_security source)
+_datastore = LocalProxy(lambda: current_app.extensions['security'].datastore)
+
+
+def create_test_user(email='test@test.org',
+                     password='', **kwargs):
+    """Create a user in the datastore, bypassing the registration process.
+
+    Returns the created user model object instance, with the plaintext password
+    as `user.password_plaintext`.
+    DO NOT USE THIS IN PRODUCTION. IT IS MEANT FOR TESTING PURPOSES ONLY.
+    """
+    assert flask.current_app.testing
+    encrypted_password = encrypt_password(password)
+    user = _datastore.create_user(email=email, password=encrypted_password,
+                                  **kwargs)
+    _datastore.commit()
+    user.password_plaintext = password
+    return user
+
+
+def login_user_via_view(client, email, password):
+    """Attempt to log the given user in via the 'login' view on the client.
+
+    Returns the response object.
+    """
+    return client.post(url_for_security('login'),
+                       data={'email': email, 'password': password},
+                       environ_base={'REMOTE_ADDR': '127.0.0.1'})
+    # If the REMOTE_ADDR isn't set it'll throw out a ValueError as it attempts
+    # to update the User model in the database with 'untrackable' as the new
+    # `last_login_ip`.
+
+
+def client_authenticated(client):
+    """Attempt to get the change password page with the given client object.
+
+    Returns False if redirected to login page,
+    True if the client is authenticated (and thus does not get redirected).
+
+    The client is "authenticated" if its `cookie_jar` contains a valid session
+    cookie or something. Probably it attaches its session cookie to the
+    request, as that's what flask-security's authentication decorators
+    seem to check for.
+    """
+    response = client.get(url_for_security('change_password'),
+                          follow_redirects=True)
+
+    return (url_for_security('login') not in response.data.decode('utf-8') and
```

Comment:
why do you need to `.decode` to response data? it should be "cheaper" to encode the url if needed